app/routes/_auth.signup.tsx

```diff
@@ -14,32 +17,56 @@ export async function action({ request }: ActionArgs) {
   const password = form.get('password');
   const confirmPassword = form.get('confirmPassword');
 
-  if (!name || !email || !password || !confirmPassword) {
-    return { status: 400, message: 'Missing required fields' };
-  }
+  const SignUpForm = z.object({
+    name: z
+      .string()
+      .trim()
+      .min(1, 'Name is required.')
+      .max(191, 'Name cannot exceed 191 max length.')
+      .transform((name) => name.replace(/\s+/g, ' ')),
+    email: z
+      .string()
+      .toLowerCase()
+      .trim()
+      .email('Invalid email.')
+      .refine(
+        async (email) => !(await db.user.findUnique({ where: { email } })),
+        'Email already in use. Is this you? Please log in instead.',
+      ),
+    password: z
+      .string()
+      .min(8, 'Password must be 8 or more characters.')
+      .max(191, 'Password cannot exceed 191 max length.'),
+    confirmPassword: z.string().refine((confirmPassword) => confirmPassword === password, 'Password must match.'),
+  });
 
-  if (password !== confirmPassword) {
-    return { status: 400, message: 'Passwords do not match' };
-  }
+  const signUpForm = await SignUpForm.safeParseAsync({ name, email, password, confirmPassword });
 
-  const cleanEmail: string = email.toString().toLowerCase().trim();
-  const user = await db.user.findUnique({ where: { email: cleanEmail } });
+  const actionData = {
+    success: false,
+    fields: { email, name },
+    fieldErrors: null,
+    formError: null,
+  };
 
-  if (user) {
-    return { status: 400, message: 'Email already in use. Is this you? Please log in instead.' };
+  if (!signUpForm.success) {
+    return badRequest({
+      ...actionData,
+      fieldErrors: signUpForm.error.format(),
+    });
   }
 
   const newUser = await db.user.create({
     data: {
-      name: name.toString(),
-      email: cleanEmail,
+      name: signUpForm.data.name,
+      email: signUpForm.data.email,
```

Comment:
We are using data from the parsed zod object here, which is already cleaned and ready to be added to our db.